lizard_auth_server/backends.py

```diff
@@ -0,0 +1,127 @@
+(BSD licensed)
+
+"""
+from boto3.exceptions import Boto3Error
+from botocore.exceptions import ClientError
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.contrib.auth.backends import ModelBackend
+from django.utils.six import iteritems
+from warrant import Cognito
+
+import abc
+
+
+def cognito_to_dict(attr_list, mapping):
+    user_attrs = dict()
+    for i in attr_list:
+        name = mapping.get(i.get("Name"))
+        if name:
+            value = i.get("Value")
+            user_attrs[name] = value
+    return user_attrs
+
+
+class CognitoUser(Cognito):
+    user_class = get_user_model()
+    # Mapping of Cognito User attribute name to Django User attribute name
+    COGNITO_ATTR_MAPPING = getattr(
+        settings,
+        "COGNITO_ATTR_MAPPING",
+        {
+            "email": "email",
+            "given_name": "first_name",
+            "family_name": "last_name",
+        },
+    )
+
+    def get_user_obj(self, username=None, attribute_list=[], metadata={}, attr_map={}):
+        user_attrs = cognito_to_dict(attribute_list, CognitoUser.COGNITO_ATTR_MAPPING)
+        django_fields = [f.name for f in CognitoUser.user_class._meta.get_fields()]
+        extra_attrs = {}
+        for k, v in user_attrs.items():
+            if k not in django_fields:
+                extra_attrs.update({k: user_attrs.pop(k, None)})
+        if getattr(settings, "COGNITO_CREATE_UNKNOWN_USERS", True):
+            user, created = CognitoUser.user_class.objects.update_or_create(
+                username=username, defaults=user_attrs
+            )
+        else:
+            try:
+                user = CognitoUser.user_class.objects.get(username=username)
+                for k, v in iteritems(user_attrs):
+                    setattr(user, k, v)
+                user.save()
+            except CognitoUser.user_class.DoesNotExist:
+                user = None
+        if user:
+            for k, v in extra_attrs.items():
+                setattr(user, k, v)
+        return user
+
+
+class AbstractCognitoBackend(ModelBackend):
+    __metaclass__ = abc.ABCMeta
+
+    UNAUTHORIZED_ERROR_CODE = "NotAuthorizedException"
+
+    USER_NOT_FOUND_ERROR_CODE = "UserNotFoundException"
+
+    COGNITO_USER_CLASS = CognitoUser
+
+    @abc.abstractmethod
+    def authenticate(self, username=None, password=None):
+        """
+        Authenticate a Cognito User
+        :param username: Cognito username
+        :param password: Cognito password
+        :return: returns User instance of AUTH_USER_MODEL or None
+        """
+        cognito_user = CognitoUser(
+            settings.COGNITO_USER_POOL_ID,
+            settings.COGNITO_APP_ID,
+            access_key=getattr(settings, "AWS_ACCESS_KEY_ID", None),
+            secret_key=getattr(settings, "AWS_SECRET_ACCESS_KEY", None),
+            username=username,
+        )
+        try:
+            cognito_user.authenticate(password)
```

Comment:
`authenticate` gebruikt de USER_SRP_AUTH flow.

Dit is niet compatibel is met de user-migration trigger. Zie:

https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-import-using-lambda.html

"Your app sends the username and password to Amazon Cognito. If your app has a native sign-in UI and uses the Cognito Identity Provider SDK, your app must use the USER_PASSWORD_AUTH flow, in which the SDK sends the password to the server (your app must not use the default USER_SRP_AUTH flow since the SDK does not send the password to the server in the SRP authentication flow). The USER_PASSWORD_AUTH flow is enabled by setting AuthenticationDetails.authenticationType to "USER_PASSWORD". "

In https://github.com/capless/warrant/blob/master/warrant/__init__.py#L368 zie ik dat `.admin_authenticate` wel een niet-SRP flow gebruikt, maar nog steeds niet te goeie. Wellicht moet hier nog wat code gekopieerd worden uit warrant zodat we de USER_PASSWORD_FLOW gebruiken.

Reply:
admin_authenticate() deed het 'em. ADMIN_USER_PASSWORD_AUTH gebruikt.